lib/ddtrace/sampling/rule_sampler.rb

```diff
@@ -0,0 +1,56 @@
+require 'ddtrace/ext/priority'
+
+require 'ddtrace/ext/sampling'
+require 'ddtrace/sampler'
+
+module Datadog
+  module Sampling
+    # TODO:
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :priority_sampler
+
+      def initialize(rules, rate_limiter, priority_sampler = Datadog::RateByServiceSampler.new)
+        @rules = rules
+        @rate_limiter = rate_limiter
+        @priority_sampler = priority_sampler
+      end
+
+      def sample?(span)
+        sampled, _ = sample_span(span) { |s| return @priority_sampler.sample?(s) }
+        sampled
+      end
+
+      def sample!(span)
+        sampled, sample_rate = sample_span(span) { |s| return @priority_sampler.sample!(s) }
+
+        # Set metrics regardless of sampling outcome
+        set_metrics(span, sample_rate, rate_limiter.effective_rate)
```

Comment:
we only want to set the rate if the sampler had a hand in making the decision.

e.g. if the rate limiter doesn't get evaluated, then we don't set the metric for the rate limiter.

Can think of this as sort of debugging for why the span was sampled or not. e.g. if the rule decides not to sample, then rate limiter didn't play a role in making that decision, we can see that by the rate limit psr metric not being set.

Reply:
Got it, I'll address to match this requirement.

Reply:
I've made changes to only set rule based metrics when rules apply to a trace.
The same goes to the rate limiter: only set metrics when the limiter was evaluated.